File: hashcash/hashcash.swift
import Foundation

// Algorithm: https://en.wikipedia.org/wiki/Hashcash
// Example: 1:20:1303030600:anni@cypherspace.org::McMybZIhxKXu57jd:ckvi

let email = "anni@cypherspace.org"
let defaultBits = 12
let formatter = DateFormatter()
formatter.dateFormat = "yyMMddHHmm"

func base64(numbers: [UInt8]) -> String {
  var data = Data()
  data.append(contentsOf: numbers)
  return data.base64EncodedString()
}

func base64(number: Int) -> String {
  var numbers: [UInt8] = []
  var n = number
  while n > 0 {
    numbers.append(UInt8(n & 0xff))
    n >>= 8
  }
  return base64(numbers: numbers)
}

func random(bytes: Int) -> String {
  var numbers = [UInt8]()
  for _ in 0..<bytes {
    numbers.append(UInt8.random(in:0..<UInt8.max))
  }
  return base64(numbers:numbers)
}

class Hashcash: NSObject {
  var ver = 1
  var bits = defaultBits
  var date = ""
  var resource = ""
  var ext = ""
  var rand = ""
  var counter = ""

  init(resource: String) {
    self.date = formatter.string(from: Date())
    self.resource = resource
    self.rand = random(bytes:12)
  }

  init?(string: String) {
    let parts = string.components(separatedBy:":")
    if parts.count == 7 {
      self.ver = Int(parts[0]) ?? 0
      self.bits = Int(parts[1]) ?? 0
      self.date = parts[2]
      self.resource = parts[3]
      self.ext = parts[4]
      self.rand = parts[5]
      self.counter = parts[6]
    } else {
      return nil
    }
  }

  func proofOfWork() {
    let firstPart = self.descriptionNoCounter
    let max = 2 << (32 - self.bits - 1)
    for i in 0..<1000000 {
      let counter = base64(number: i)
      let value = "\(firstPart)\(counter)"
      let hash = SHA1.hash(from: value)![0]
      if hash < max {
        self.counter = counter
        return
      }
    }
    print("Did not find hash less than \(max)!")
  }

  static func generate(resource: String) -> String {
    let hashcash = Hashcash(resource: resource)
    hashcash.proofOfWork()
    return hashcash.description
  }
  
  static func validate(string: String, resource: String) -> Bool {
    guard let hashcash = Hashcash(string: string) else {
      print("Couldn't parse string: \(string)")
      return false
    }
    // version is correct
    if hashcash.ver != 1 {
      print("Unknown version: \(hashcash.ver)")
      return false
    }
    // date is valid format
    guard let date = formatter.date(from: hashcash.date) else {
      print("Invalid date: \(hashcash.date)")
      return false
    }
    // date is within one day
    let age = date.timeIntervalSinceNow
    if age > 86400 || age < -86400 {
      print("Timestamp out of range: \(age) seconds")
      return false
    }
    // resource is the same
    if resource != hashcash.resource {
      print("Resource mismatch: \(resource)/\(hashcash.resource)")
      return false
    }
    // random data is correct length
    if hashcash.rand.count < 16 {
      print("Random too short: \(hashcash.rand.count)")
      return false
    }
    // SHA1 hash has correct number of leading zero bits
    let hash = SHA1.hash(from: string)![0]
    let max = 2 << (32 - hashcash.bits - 1)
    if hash > max {
      print("hash \(hash) is greater than max \(max)")
      return false
    }
    return true
  }

  override var description: String {
    return [String(self.ver), String(self.bits), 
      self.date, self.resource, self.ext, self.rand,
      self.counter].joined(separator:":")
  }

  var descriptionNoCounter: String {
    return [String(self.ver), String(self.bits), 
      self.date, self.resource, self.ext, self.rand,
      ""].joined(separator:":")
  }
}

let value = Hashcash.generate(resource: email)
print(value)
if Hashcash.validate(string: value, resource: email) {
  print("Validated")
}

infix operator <<< : BitwiseShiftPrecedence
private func <<< (lhs:UInt32, rhs:UInt32) -> UInt32 {
    return lhs << rhs | lhs >> (32-rhs)
}

public struct SHA1 {
  private static let CHUNKSIZE=80
  private static let h0:UInt32 = 0x67452301
  private static let h1:UInt32 = 0xEFCDAB89
  private static let h2:UInt32 = 0x98BADCFE
  private static let h3:UInt32 = 0x10325476
  private static let h4:UInt32 = 0xC3D2E1F0

  private struct context {
    var h:[UInt32]=[SHA1.h0,SHA1.h1,SHA1.h2,SHA1.h3,SHA1.h4]

    mutating func process(chunk:inout ContiguousArray<UInt32>) {
      for i in 0..<16 {
        chunk[i] = chunk[i].bigEndian
      }
      for i in 16...79 {
        chunk[i] = (chunk[i-3] ^ chunk[i-8] ^ chunk[i-14] ^ chunk[i-16]) <<< 1
      }

      var a,b,c,d,e,f,k,temp:UInt32
      a=h[0]; b=h[1]; c=h[2]; d=h[3]; e=h[4]
      f=0x0; k=0x0

      for i in 0...79 {
        switch i {
        case 0...19:
          f = (b & c) | ((~b) & d)
          k = 0x5A827999
        case 20...39:
          f = b ^ c ^ d
          k = 0x6ED9EBA1
        case 40...59:
          f = (b & c) | (b & d) | (c & d)
          k = 0x8F1BBCDC
        case 60...79:
          f = b ^ c ^ d
          k = 0xCA62C1D6
        default: break
        }
        temp = a <<< 5 &+ f &+ e &+ k &+ chunk[i]
        e = d
        d = c
        c = b <<< 30
        b = a
        a = temp
      }

      h[0] = h[0] &+ a
      h[1] = h[1] &+ b
      h[2] = h[2] &+ c
      h[3] = h[3] &+ d
      h[4] = h[4] &+ e
    }
  }

  private static func process(data: inout Data) -> SHA1.context? {
    var context=SHA1.context()
    var w = ContiguousArray<UInt32>(repeating: 0x00000000, count: CHUNKSIZE)
    let ml=data.count << 3
    var range = 0..<64

    while data.count >= range.upperBound {
      w.withUnsafeMutableBufferPointer{ dest in
        data.copyBytes(to: dest, from: range)
      }
      context.process(chunk: &w)
      range = range.upperBound..<range.upperBound+64
    }

    w = ContiguousArray<UInt32>(repeating: 0x00000000, count: CHUNKSIZE)
    range = range.lowerBound..<data.count
    w.withUnsafeMutableBufferPointer{ dest in
      data.copyBytes(to: dest, from: range)
    }
    let bytetochange=range.count % 4
    let shift = UInt32(bytetochange * 8)
    w[range.count/4] |= 0x80 << shift
    if range.count+1 > 56 {
      context.process(chunk: &w)
      w = ContiguousArray<UInt32>(repeating: 0x00000000, count: CHUNKSIZE)
    }

    w[15] = UInt32(ml).bigEndian
    context.process(chunk: &w)
    return context
  }

  public static func hash(from str:String) -> [Int]? {
    guard var data = str.data(using: .utf8) else { return nil }
    return process(data: &data)?.h.map{Int($0)}
  }
}

File: hashcash/hashcash.ts
// SHA1
// https://gist.github.com/ducksoupdev/bbfcf8b9cee688b97865
module sha1 {
  var POW_2_24 = Math.pow(2, 24);
  var POW_2_32 = Math.pow(2, 32);

  function hex(n: number): string {
    var s = "",
      v: number;
    for (var i = 7; i >= 0; --i)
    {
      v = (n >>> (i << 2)) & 0xF;
      s += v.toString(16);
    }
    return s;
  }

  function lrot(n: number, bits: number): number {
    return ((n << bits) | (n >>> (32 - bits)));
  }

  class Uint32ArrayBigEndian {
    bytes: Uint8Array;
    constructor(length: number) {
      this.bytes = new Uint8Array(length << 2);
    }
    get(index: number): number {
      index <<= 2;
      return (this.bytes[index] * POW_2_24)
        + ((this.bytes[index + 1] << 16)
        | (this.bytes[index + 2] << 8)
        | this.bytes[index + 3]);
    }
    set(index: number, value: number) {
      var high = Math.floor(value / POW_2_24),
        rest = value - (high * POW_2_24);
      index <<= 2;
      this.bytes[index] = high;
      this.bytes[index + 1] = rest >> 16;
      this.bytes[index + 2] = (rest >> 8) & 0xFF;
      this.bytes[index + 3] = rest & 0xFF;
    }
  }

  function string2ArrayBuffer(s: string): ArrayBuffer {
    s = s.replace(/[\u0080-\u07ff]/g,
      function(c: string) {
        var code = c.charCodeAt(0);
        return String.fromCharCode(0xC0 | code >> 6, 0x80 | code & 0x3F);
      });
    s = s.replace(/[\u0080-\uffff]/g,
      function(c: string) {
        var code = c.charCodeAt(0);
        return String.fromCharCode(0xE0 | code >> 12, 0x80 | code >> 6 & 0x3F, 0x80 | code & 0x3F);
      });
    var n = s.length,
      array = new Uint8Array(n);
    for (var i = 0; i < n; ++i) {
      array[i] = s.charCodeAt(i);
    }
    return array.buffer;
  }

  export function hash(bufferOrString: any): string {
    var source: ArrayBuffer;
    if (bufferOrString instanceof ArrayBuffer) {
      source = <ArrayBuffer> bufferOrString;
    } else {
      source = string2ArrayBuffer(String(bufferOrString));
    }

    var h0 = 0x67452301,
      h1 = 0xEFCDAB89,
      h2 = 0x98BADCFE,
      h3 = 0x10325476,
      h4 = 0xC3D2E1F0,
      i: number,
      sbytes = source.byteLength,
      sbits = sbytes << 3,
      minbits = sbits + 65,
      bits = Math.ceil(minbits / 512) << 9,
      bytes = bits >>> 3,
      slen = bytes >>> 2,
      s = new Uint32ArrayBigEndian(slen),
      s8 = s.bytes,
      j: number,
      w = new Uint32Array(80),
      sourceArray = new Uint8Array(source);
    for (i = 0; i < sbytes; ++i) {
      s8[i] = sourceArray[i];
    }
    s8[sbytes] = 0x80;
    s.set(slen - 2, Math.floor(sbits / POW_2_32));
    s.set(slen - 1, sbits & 0xFFFFFFFF);
    for (i = 0; i < slen; i += 16) {
      for (j = 0; j < 16; ++j) {
        w[j] = s.get(i + j);
      }
      for ( ; j < 80; ++j) {
        w[j] = lrot(w[j - 3] ^ w[j - 8] ^ w[j - 14] ^ w[j - 16], 1);
      }
      var a = h0,
        b = h1,
        c = h2,
        d = h3,
        e = h4,
        f: number,
        k: number,
        temp: number;
      for (j = 0; j < 80; ++j) {
        if (j < 20) {
          f = (b & c) | ((~b) & d);
          k = 0x5A827999;
        } else if (j < 40) {
          f = b ^ c ^ d;
          k = 0x6ED9EBA1;
        } else if (j < 60) {
          f = (b & c) ^ (b & d) ^ (c & d);
          k = 0x8F1BBCDC;
        } else {
          f = b ^ c ^ d;
          k = 0xCA62C1D6;
        }

        temp = (lrot(a, 5) + f + e + k + w[j]) & 0xFFFFFFFF;
        e = d;
        d = c;
        c = lrot(b, 30);
        b = a;
        a = temp;
      }
      h0 = (h0 + a) & 0xFFFFFFFF;
      h1 = (h1 + b) & 0xFFFFFFFF;
      h2 = (h2 + c) & 0xFFFFFFFF;
      h3 = (h3 + d) & 0xFFFFFFFF;
      h4 = (h4 + e) & 0xFFFFFFFF;
    }
    return hex(h0) + hex(h1) + hex(h2) + hex(h3) + hex(h4);
  }
}

// BASE64

function bin2String(numbers: number[]): string {
  return numbers.map((i) => String.fromCharCode(i)).join("");
}

const b64ch = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=';
const b64chs = Array.prototype.slice.call(b64ch);

function btoaPolyfill(bin: string): string {
    // console.log('polyfilled');
    let u32, c0, c1, c2, asc = '';
    const pad = bin.length % 3;
    for (let i = 0; i < bin.length;) {
        if ((c0 = bin.charCodeAt(i++)) > 255 ||
            (c1 = bin.charCodeAt(i++)) > 255 ||
            (c2 = bin.charCodeAt(i++)) > 255)
            throw new TypeError('invalid character found');
        u32 = (c0 << 16) | (c1 << 8) | c2;
        asc += b64chs[u32 >> 18 & 63]
            + b64chs[u32 >> 12 & 63]
            + b64chs[u32 >> 6 & 63]
            + b64chs[u32 & 63];
    }
    return pad ? asc.slice(0, pad - 3) + "===".substring(pad) : asc;
};

// UTILS

function byteArray(n: number): number[] {
  let numbers: number[] = [];
  while (n > 255) {
    numbers.push(Math.floor(n / 256));
    n = n % 256;
  }
  numbers.push(n);
  return numbers;
}

function randomNumbers(count: number): number[] {
  return [...Array(count)].map((_) => 
     Math.floor((Math.random() * 256) + 1));
}

function makeDate() {
  let today = new Date();
  let iso = today.toISOString();
  return iso.substring(2,4) + 
    iso.substring(5,7) + 
    iso.substring(8,10) + 
    iso.substring(11,13) + 
    iso.substring(14,16);
}

function makeRand() {
  return btoaPolyfill(bin2String(randomNumbers(12)));
}

function encodeNumber(n: number) {
  return btoaPolyfill(bin2String(byteArray(n)));
}

function zeroPrefix(bytes: number): string {
  return [...Array(Math.floor(bytes / 4))].map((_) => "0").join("");
}

// HASHCASH

let email = "anni@cypherspace.org";
let defaultBits = 12;

interface Hashcash {
  ver: number;
  bits: number;
  date: string;
  resource: string;
  ext: string;
  rand: string;
  counter: string;
}

function join(h: Hashcash, includeCounter: boolean = true): string {
  let vals = [h.ver, h.bits, h.date, h.resource, h.ext, h.rand, includeCounter ? h.counter : ""];
  return vals.join(":");
}

function parse(value: string): Hashcash {
  let values = value.split(":");
  let hashcash = {
    ver: parseInt(values[0]),
    bits: parseInt(values[1]),
    date: values[2],
    resource: values[3],
    ext: values[4],
    rand: values[5],
    counter: values[6]
  };
  return hashcash;
}

function findCounter(hashcash: Hashcash) {
  let prefix = join(hashcash, false);
  let zeroes = zeroPrefix(hashcash.bits);
  let found = false;
  for (let i = 0; !found; i++) {
    let counter = encodeNumber(i);
    let hash = sha1.hash(prefix + counter);
    if (hash.startsWith(zeroes)) {
      hashcash.counter = counter;
      found = true;
    }
  }
}

function generate(resource: string): string {
  let hashcash = {
    ver: 1,
    bits: defaultBits,
    date: makeDate(),
    resource,
    ext: "",
    rand: makeRand(),
    counter: ""
  };

  findCounter(hashcash);
  return join(hashcash);
}

function validate(value: string, resource: string): boolean {
  let hashcash = parse(value);

  // version is correct
  if (hashcash.ver != 1) {
    console.log("Unknown version");
    return false;
  }
  // date is the same day
  if (!hashcash.date.startsWith(makeDate().substring(0,6))) {
    console.log("Not the same day");
    return false;
  }
  // resource is the same
  if (resource != hashcash.resource) {
    console.log("Resource not the same");
    return false;
  }
  // random data is correct length
  if (hashcash.rand.length < 16) {
    console.log("Random too short");
    return false;
  }
  // SHA1 hash has correct number of leading zeros
  let hash = sha1.hash(value);
  let zeroes = zeroPrefix(hashcash.bits);
  if (!hash.startsWith(zeroes)) {
    console.log("Hash doesn't start with zeroes");
    return false;
  }
  
  return true;
}

let value = generate(email);
console.log(value);
if (validate(value, email)) {
  console.log("Validated");
}
